### ISSUE TITLE
kubedock high CPU usage if pods stuck in CrashLoopBackOff

### ISSUE THREAD

**Issue by jkroepke** [joyrex2001/kubedock#2]
Hi,

today, I found this amazing project to close the gab between testcontainers and gitlab ci which runs on kubernetes, too. Thanks for this awesome work!

Before I start to integrate kubedock, I test it locally. While initial tests are fine, running https://github.com/rieckpil/blog-tutorials/tree/master/spring-boot-integration-tests-testcontainers results into a high CPU usage.

![image](https://user-images.githubusercontent.com/1560587/139117570-ea228e2b-e260-433a-8003-0ab3d1ef092e.png)

Logs:

```
jkr@joe-nb ~ % ~/Downloads/kubedock server --port-forward
I1027 19:35:11.748402   84001 main.go:26] kubedock 0.7.0 (20211008-105904)
I1027 19:35:11.749336   84001 main.go:95] kubernetes config: namespace=vrp-testcontainers-kubernetes, initimage=joyrex2001/kubedock:0.7.0, ready timeout=1m0s
I1027 19:35:11.749668   84001 main.go:117] reaper started with max container age 1h0m0s
I1027 19:35:11.749770   84001 main.go:68] port-forwarding services to 127.0.0.1
I1027 19:35:11.749885   84001 main.go:100] default image pull policy: ifnotpresent
I1027 19:35:11.749926   84001 main.go:102] using namespace: vrp-testcontainers-kubernetes
I1027 19:35:11.750065   84001 main.go:35] api server started listening on :2475
[GIN] 2021/10/27 - 19:35:20 | 200 |     123.807µs |       127.0.0.1 | GET      "/info"
[GIN] 2021/10/27 - 19:35:20 | 200 |      29.519µs |       127.0.0.1 | GET      "/info"
[GIN] 2021/10/27 - 19:35:20 | 200 |      28.059µs |       127.0.0.1 | GET      "/version"
[GIN] 2021/10/27 - 19:35:20 | 200 |      75.252µs |       127.0.0.1 | GET      "/images/json"
[GIN] 2021/10/27 - 19:35:20 | 200 |        78.5µs |       127.0.0.1 | GET      "/images/jboss/keycloak:11.0.0/json"
[GIN] 2021/10/27 - 19:35:20 | 201 |     394.528µs |       127.0.0.1 | POST     "/containers/create"
[GIN] 2021/10/27 - 19:35:35 | 204 | 14.567361081s |       127.0.0.1 | POST     "/containers/815f8424ab54d5c94987109c3049e005dd620b1c2904f2f8aea3580ee2999369/start"
I1027 19:35:35.422243   84001 portforward.go:42] start port-forward 34468->8080
[GIN] 2021/10/27 - 19:35:35 | 200 |     120.223µs |       127.0.0.1 | GET      "/containers/815f8424ab54d5c94987109c3049e005dd620b1c2904f2f8aea3580ee2999369/json"
[GIN] 2021/10/27 - 19:35:35 | 200 |     124.327µs |       127.0.0.1 | GET      "/containers/815f8424ab54d5c94987109c3049e005dd620b1c2904f2f8aea3580ee2999369/json"
E1027 19:35:36.603462   84001 portforward.go:400] an error occurred forwarding 34468 -> 8080: error forwarding port 8080 to pod 97b9789ffbcc98354c10c90815c06897e999e1c69005fcfbec26c9e870795bb1, uid : exit status 1: 2021/10/27 19:35:36 socat[60751] E connect(5, AF=2 127.0.0.1:8080, 16): Connection refused
E1027 19:35:36.676557   84001 portforward.go:400] an error occurred forwarding 34468 -> 8080: error forwarding port 8080 to pod 97b9789ffbcc98354c10c90815c06897e999e1c69005fcfbec26c9e870795bb1, uid : exit status 1: 2021/10/27 19:35:36 socat[60758] E connect(5, AF=2 127.0.0.1:8080, 16): Connection refused
E1027 19:35:37.762316   84001 portforward.go:400] an error occurred forwarding 34468 -> 8080: error forwarding port 8080 to pod 97b9789ffbcc98354c10c90815c06897e999e1c69005fcfbec26c9e870795bb1, uid : exit status 1: 2021/10/27 19:35:37 socat[60924] E connect(5, AF=2 127.0.0.1:8080, 16): Connection refused
E1027 19:35:37.832716   84001 portforward.go:400] an error occurred forwarding 34468 -> 8080: error forwarding port 8080 to pod 97b9789ffbcc98354c10c90815c06897e999e1c69005fcfbec26c9e870795bb1, uid : exit status 1: 2021/10/27 19:35:37 socat[60931] E connect(5, AF=2 127.0.0.1:8080, 16): Connection refused
E1027 19:35:38.912798   84001 portforward.go:400] an error occurred forwarding 34468 -> 8080: error forwarding port 8080 to pod 97b9789ffbcc98354c10c90815c06897e999e1c69005fcfbec26c9e870795bb1, uid : exit status 1: 2021/10/27 19:35:38 socat[61039] E connect(5, AF=2 127.0.0.1:8080, 16): Connection refused
E1027 19:35:38.987719   84001 portforward.go:400] an error occurred forwarding 34468 -> 8080: error forwarding port 8080 to pod 97b9789ffbcc98354c10c90815c06897e999e1c69005fcfbec26c9e870795bb1, uid : exit status 1: 2021/10/27 19:35:39 socat[61046] E connect(5, AF=2 127.0.0.1:8080, 16): Connection refused
E1027 19:35:40.076636   84001 portforward.go:400] an error occurred forwarding 34468 -> 8080: error forwarding port 8080 to pod 97b9789ffbcc98354c10c90815c06897e999e1c69005fcfbec26c9e870795bb1, uid : exit status 1: 2021/10/27 19:35:40 socat[61095] E connect(5, AF=2 127.0.0.1:8080, 16): Connection refused
E1027 19:35:40.158013   84001 portforward.go:400] an error occurred forwarding 34468 -> 8080: error forwarding port 8080 to pod 97b9789ffbcc98354c10c90815c06897e999e1c69005fcfbec26c9e870795bb1, uid : exit status 1: 2021/10/27 19:35:40 socat[61107] E connect(5, AF=2 127.0.0.1:8080, 16): Connection refused
[GIN] 2021/10/27 - 19:35:53 | 201 |      84.187µs |       127.0.0.1 | POST     "/containers/815f8424ab54d5c94987109c3049e005dd620b1c2904f2f8aea3580ee2999369/exec"
[GIN] 2021/10/27 - 19:35:54 | 200 |  317.038459ms |       127.0.0.1 | POST     "/exec/808a7da1789efc9f5e8a0b8bdf5b8ca44843e0dddcaeed5ab7e0a331870c2029/start"
[GIN] 2021/10/27 - 19:35:54 | 200 |      68.388µs |       127.0.0.1 | GET      "/exec/808a7da1789efc9f5e8a0b8bdf5b8ca44843e0dddcaeed5ab7e0a331870c2029/json"
[GIN] 2021/10/27 - 19:35:54 | 200 |      83.432µs |       127.0.0.1 | GET      "/containers/815f8424ab54d5c94987109c3049e005dd620b1c2904f2f8aea3580ee2999369/json"
I1027 19:35:54.033603   84001 containers.go:217] ignoring signal
[GIN] 2021/10/27 - 19:35:54 | 204 |      42.978µs |       127.0.0.1 | POST     "/containers/815f8424ab54d5c94987109c3049e005dd620b1c2904f2f8aea3580ee2999369/kill"
[GIN] 2021/10/27 - 19:35:54 | 200 |      90.172µs |       127.0.0.1 | GET      "/containers/815f8424ab54d5c94987109c3049e005dd620b1c2904f2f8aea3580ee2999369/json"
[GIN] 2021/10/27 - 19:35:54 | 204 |  253.744279ms |       127.0.0.1 | DELETE   "/containers/815f8424ab54d5c94987109c3049e005dd620b1c2904f2f8aea3580ee2999369?v=true&force=true"
[GIN] 2021/10/27 - 19:35:54 | 200 |     105.337µs |       127.0.0.1 | GET      "/images/postgres:12/json"
[GIN] 2021/10/27 - 19:35:54 | 201 |     175.279µs |       127.0.0.1 | POST     "/containers/create"
I1027 19:36:19.621781   84001 portforward.go:42] start port-forward 47785->5432
[GIN] 2021/10/27 - 19:36:19 | 204 | 25.308206574s |       127.0.0.1 | POST     "/containers/a1f2506d65208460cd20342ff35c4db80eea8752700e1055f123a18148856935/start"
[GIN] 2021/10/27 - 19:36:19 | 200 |     129.701µs |       127.0.0.1 | GET      "/containers/a1f2506d65208460cd20342ff35c4db80eea8752700e1055f123a18148856935/json"
[GIN] 2021/10/27 - 19:36:19 | 200 |      81.498µs |       127.0.0.1 | GET      "/containers/a1f2506d65208460cd20342ff35c4db80eea8752700e1055f123a18148856935/json"
[GIN] 2021/10/27 - 19:37:19 | 200 |     317.497µs |       127.0.0.1 | GET      "/containers/a1f2506d65208460cd20342ff35c4db80eea8752700e1055f123a18148856935/json"
[GIN] 2021/10/27 - 19:37:19 | 200 |   74.017123ms |       127.0.0.1 | GET      "/containers/a1f2506d65208460cd20342ff35c4db80eea8752700e1055f123a18148856935/logs?stdout=true&stderr=true&since=0"
[GIN] 2021/10/27 - 19:37:19 | 200 |    58.57439ms |       127.0.0.1 | GET      "/containers/a1f2506d65208460cd20342ff35c4db80eea8752700e1055f123a18148856935/logs?stdout=true&stderr=true&since=0"
[GIN] 2021/10/27 - 19:37:19 | 201 |     190.866µs |       127.0.0.1 | POST     "/containers/create"
[GIN] 2021/10/27 - 19:37:23 | 204 |  3.277132909s |       127.0.0.1 | POST     "/containers/2d2f2db70a487e9b10f54bef0580359f9d69a39b99e0057a1ea4492a7e78a2e7/start"
I1027 19:37:23.216054   84001 portforward.go:42] start port-forward 62630->5432
[GIN] 2021/10/27 - 19:37:23 | 200 |      95.659µs |       127.0.0.1 | GET      "/containers/2d2f2db70a487e9b10f54bef0580359f9d69a39b99e0057a1ea4492a7e78a2e7/json"
[GIN] 2021/10/27 - 19:37:23 | 200 |      86.154µs |       127.0.0.1 | GET      "/containers/2d2f2db70a487e9b10f54bef0580359f9d69a39b99e0057a1ea4492a7e78a2e7/json"
[GIN] 2021/10/27 - 19:38:23 | 200 |     116.501µs |       127.0.0.1 | GET      "/containers/2d2f2db70a487e9b10f54bef0580359f9d69a39b99e0057a1ea4492a7e78a2e7/json"
[GIN] 2021/10/27 - 19:38:23 | 200 |   69.953111ms |       127.0.0.1 | GET      "/containers/2d2f2db70a487e9b10f54bef0580359f9d69a39b99e0057a1ea4492a7e78a2e7/logs?stdout=true&stderr=true&since=0"
[GIN] 2021/10/27 - 19:38:23 | 200 |   61.136457ms |       127.0.0.1 | GET      "/containers/2d2f2db70a487e9b10f54bef0580359f9d69a39b99e0057a1ea4492a7e78a2e7/logs?stdout=true&stderr=true&since=0"
[GIN] 2021/10/27 - 19:38:23 | 201 |     285.397µs |       127.0.0.1 | POST     "/containers/create"
[GIN] 2021/10/27 - 19:38:28 | 204 |  4.293655856s |       127.0.0.1 | POST     "/containers/169389ceb3c51085fc7b0c14a4f99972369a4fd039b98b8afe41da50b75830d6/start"
I1027 19:38:28.117308   84001 portforward.go:42] start port-forward 62426->5432
[GIN] 2021/10/27 - 19:38:28 | 200 |     203.466µs |       127.0.0.1 | GET      "/containers/169389ceb3c51085fc7b0c14a4f99972369a4fd039b98b8afe41da50b75830d6/json"
[GIN] 2021/10/27 - 19:38:28 | 200 |     164.063µs |       127.0.0.1 | GET      "/containers/169389ceb3c51085fc7b0c14a4f99972369a4fd039b98b8afe41da50b75830d6/json"
[GIN] 2021/10/27 - 19:39:28 | 200 |     146.367µs |       127.0.0.1 | GET      "/containers/169389ceb3c51085fc7b0c14a4f99972369a4fd039b98b8afe41da50b75830d6/json"
[GIN] 2021/10/27 - 19:39:28 | 200 |   66.033642ms |       127.0.0.1 | GET      "/containers/169389ceb3c51085fc7b0c14a4f99972369a4fd039b98b8afe41da50b75830d6/logs?stdout=true&stderr=true&since=0"
[GIN] 2021/10/27 - 19:39:28 | 200 |   62.444152ms |       127.0.0.1 | GET      "/containers/169389ceb3c51085fc7b0c14a4f99972369a4fd039b98b8afe41da50b75830d6/logs?stdout=true&stderr=true&since=0"
```

Kubedock: 0.7.0
OS: Mac OS
Kubernetes: Openshift 3.11

```
jkr@joe-nb ~ % kubectl get all -l kubedock=true
NAME                                READY   STATUS             RESTARTS   AGE
pod/169389ceb3c5-85446cf64b-66qpb   0/1     CrashLoopBackOff   5          3m
pod/2d2f2db70a48-65bdb6dbb7-67rj8   0/1     CrashLoopBackOff   5          4m
pod/a1f2506d6520-c5c57c8b6-x5b65    0/1     CrashLoopBackOff   5          5m

NAME                      TYPE        CLUSTER-IP      EXTERNAL-IP   PORT(S)              AGE
service/kd-169389ceb3c5   ClusterIP   172.30.173.82   <none>        5432/TCP,62426/TCP   3m
service/kd-2d2f2db70a48   ClusterIP   172.30.227.70   <none>        5432/TCP,62630/TCP   4m
service/kd-a1f2506d6520   ClusterIP   172.30.71.160   <none>        5432/TCP,47785/TCP   5m

NAME                           DESIRED   CURRENT   UP-TO-DATE   AVAILABLE   AGE
deployment.apps/169389ceb3c5   1         1         1            0           3m
deployment.apps/2d2f2db70a48   1         1         1            0           4m
deployment.apps/a1f2506d6520   1         1         1            0           5m

NAME                                      DESIRED   CURRENT   READY   AGE
replicaset.apps/169389ceb3c5-85446cf64b   1         1         0       3m
replicaset.apps/2d2f2db70a48-65bdb6dbb7   1         1         0       4m
replicaset.apps/a1f2506d6520-c5c57c8b6    1         1         0       5m
```

This reason for this could be that the pods are in `CrashLoopBackOff`. The reason, that pods crashes are known (file permissions issues), but in such cases, kubedock should not generate such a high load. The high load is persistent even after mvn clean verify is finished. Also press CRTL+C takes some time to terminate the process.

I'm able to reproduce this behaivor. If you teach me, I'm able to provide traces or profiling files. But before doing this, please ensure that such profiling files does not contain sensitive informations like the kube credentials.

**Comment by joyrex2001** [comment]
The bug was reproducible with your sleep/exit suggestion. It was continuously trying to read logs from the crashing pod which caused the high cpu load.

**Comment by joyrex2001** [comment]
Released the fix in kubedock-0.7.1.
